File: pathgame/pathgame_initial.cpp
#include <iostream>
#include <array>

using std::cout;
using std::endl;

enum class V {
    u, // unused
    f, // free
    r, // red
    g  // green
};

static const size_t X_LEN = 4;
static const size_t Y_LEN = 5;
typedef std::array<std::array<V, Y_LEN>, X_LEN> Board;
Board b = {{
    {V::f, V::u, V::u, V::u, V::g}, // 1st column, ie. b[0][y]
    {V::f, V::f, V::f, V::g, V::r}, // b[1][y] ...
    {V::r, V::u, V::r, V::u, V::g},
    {V::r, V::u, V::u, V::u, V::g}
}};

struct Point
{
    size_t x;
    size_t y;
};

void print_val(V v)
{
    switch (v)
    {
    case V::u:
        {
            cout << ' ';
            break;
        }
    case V::f:
        {
            cout << '.';
            break;
        }
    case V::r:
        {
            cout << 'R';
            break;
        }
    case V::g:
        {
            cout << 'G';
            break;
        }
    }
 }

void print()
{
    cout << "   ";
    for (size_t x = 0; x < X_LEN; ++x)
    {
        cout << x;
        cout << ' ';
    }
    cout << endl;
    for (size_t y = 0; y < Y_LEN; ++y)
    {
        cout << y;
        cout << ' ';
        for (size_t x = 0; x < X_LEN; ++x)
        {
            cout << ' ';
            print_val(b[x][y]);
        }
        cout << endl;
    }
    cout << endl;
}

void move(const Point p1, const Point p2)
{
    if (p1.x > X_LEN - 1 || p2.x > X_LEN - 1 || p1.y > Y_LEN - 1 || p2.y > Y_LEN - 1)
    {
        cout << "ERROR: move: invalid parameter" << endl;
        return;
    }

    // TODO
}

int main()
{
    print();
/*
   0 1 2 3
0  . . R R
1    .
2    . R
3    G
4  G R G G
*/

    move({2, 2}, {0, 0});
    print();
/*
   0 1 2 3
0  R . R R
1    .
2    . .
3    G
4  G R G G
*/

    move({1, 3}, {2, 2});
    print();
/*
   0 1 2 3
0  R . R R
1    .
2    . G
3    .
4  G R G G
*/

    move({1, 4}, {1, 0});
    print();
/*
   0 1 2 3
0  R R R R
1    .
2    . G
3    .
4  G . G G
*/

    move({2, 2}, {1, 4});
    print();
/*
   0 1 2 3
0  R R R R
1    .
2    . .
3    .
4  G G G G
*/

    move({0, 0}, {0, 1});
    print();

    return 0;
}

File: pathgame/pathgame_initial.py
from dataclasses import dataclass

XLEN = 4
YLEN = 5


@dataclass
class P:
    x: int
    y: int


class Board:
    def __init__(self):
        # start with empty board:
        # [
        #   [None, None, None, None, None], = 1st column, ie. self.board[0][y]
        #   [None, None, None, None, None], = 2nd column, ie. self.board[1][y] ...
        #   [None, None, None, None, None],
        #   [None, None, None, None, None]
        # ]
        self.board = [[None]*YLEN for _ in range(0, XLEN)]

    def insert(self, p, e):
        if p.x < 0 or p.y < 0 or p.x > XLEN or p.y > YLEN:
            print("add: invalid input parameters")
            return
        self.board[p.x][p.y] = e

    def print_board(self):
        print("")
        print("   ", end="")
        for x in range(XLEN):
            print(f"{x} ", end="")
        print("")
        for y in range(YLEN):
            print(f" {y} ", end="")
            for x in range(XLEN):
                p = self.board[x][y]
                if p:
                    print(f"{p} ", end="")
                else:
                    print("  ", end="")
            print("")

    def move(self, p1, p2):
        if p1.x < 0 or p1.y < 0 or p2.x < 0 or p2.y < 0 or p1.x > (XLEN - 1) or p1.y > (YLEN - 1) or p2.x > (XLEN - 1) or p2.y > (YLEN - 1):
            print("move: invalid input parameters")
            return
        if p1.x == p2.x and p1.y == p2.y:
            print(f"error: source and dest same {p1.x},{p1.y}")
            return
        # TODO


b = Board()
b.insert(P(0, 0), ".")
b.insert(P(1, 0), ".")
b.insert(P(2, 0), "R")
b.insert(P(3, 0), "R")

b.insert(P(0, 4), "G")
b.insert(P(1, 4), "R")
b.insert(P(2, 4), "G")
b.insert(P(3, 4), "G")

b.insert(P(1, 1), ".")
b.insert(P(1, 2), ".")
b.insert(P(2, 2), "R")
b.insert(P(1, 3), "G")

b.print_board()
#   0 1 2 3
# 0 . . R R
# 1   .
# 2   . R
# 3   G
# 4 G R G G

b.move(P(2, 2), P(0, 0))
b.print_board()
#   0 1 2 3
# 0 R . R R
# 1   .
# 2   . .
# 3   G
# 4 G R G G

b.move(P(1, 3), P(2, 2))
b.print_board()
#   0 1 2 3
# 0 R . R R
# 1   .
# 2   . G
# 3   .
# 4 G R G G

b.move(P(1, 4), P(1, 0))
b.print_board()
#   0 1 2 3
# 0 R R R R
# 1   .
# 2   . G
# 3   .
# 4 G . G G

b.move(P(2, 2), P(1, 4))
b.print_board()
#   0 1 2 3
# 0 R R R R
# 1   .
# 2   . .
# 3   .
# 4 G G G G
